packages/apps/plugins/plugin-script/src/components/ScriptEditor.tsx

```diff
@@ -0,0 +1,113 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { type DID } from 'iso-did/types';
+import React, { useCallback, useMemo, useState } from 'react';
+
+import { type ScriptType } from '@braneframe/types';
+import { log } from '@dxos/log';
+import { useClient } from '@dxos/react-client';
+import { createDocAccessor, getMeta, getSpace } from '@dxos/react-client/echo';
+import { useIdentity } from '@dxos/react-client/halo';
+import { useTranslation } from '@dxos/react-ui';
+import { createDataExtensions, listener } from '@dxos/react-ui-editor';
+import { mx } from '@dxos/react-ui-theme';
+
+import { Toolbar } from './Toolbar';
+import { TypescriptEditor } from './TypescriptEditor';
+import {
+  getUserFunctionUrlInMetadata,
+  publicKeyToDid,
+  setUserFunctionUrlInMetadata,
+  uploadWorkerFunction,
+} from '../edge';
+import { SCRIPT_PLUGIN } from '../meta';
+
+export type ScriptEditorProps = {
+  script: ScriptType;
+  role?: string;
+};
+
+export const ScriptEditor = ({ script, role }: ScriptEditorProps) => {
+  const { t } = useTranslation(SCRIPT_PLUGIN);
+  const client = useClient();
+  const identity = useIdentity();
+  const space = getSpace(script);
+  const extensions = useMemo(
+    () => [
+      listener({
+        onChange: (text) => {
+          if (script.source && script.source?.content !== text) {
+            script.changed = true;
+          }
+        },
+      }),
+      createDataExtensions({
+        id: script.id,
+        text: script.source && createDocAccessor(script.source, ['content']),
+        space,
+        identity,
+      }),
+    ],
+    [script, script.source, space, identity],
+  );
+  const initialValue = useMemo(() => script.source?.content, [script.source]);
+  const existingFunctionUrl = getUserFunctionUrlInMetadata(getMeta(script));
+  const [error, setError] = useState<string>();
+
+  const handleBindingChange = useCallback(
+    (binding: string) => {
+      script.binding = binding;
+    },
+    [script],
+  );
+
+  const handleDeploy = useCallback(async () => {
```

Comment:
if this gets more complicated consider making a local file hook: useDeploy, etc?